vck/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/DocumentDigestEntry.kt

```diff
@@ -0,0 +1,158 @@
+@file:UseSerializers(UrlSerializer::class)
+
+package at.asitplus.wallet.lib.data.dif
+
+import at.asitplus.KmmResult
+import at.asitplus.KmmResult.Companion.wrap
+import io.ktor.http.*
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.UseSerializers
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+
+@Serializable
+data class DocumentDigestEntry private constructor(
+    val label: String,
+    /**
+     * base64 encoded octet representation generated using hashAlgorithmOID
+     */
+    val hash: String? = null,
+    val hashAlgorithmOID: String? = null,
+    @SerialName("documentLocation_uri") //errorous in test vector - changed for test case
+    val documentLocationUri: Url? = null,
+    @SerialName("documentLocation_method")
+    val documentLocationMethod: DocumentLocationMethod? = null,
+    /**
+     * base64 encoded octet representation generated using dtbsrAlgorithmOID
+     */
+    val dtbsr: String? = null,
+    val dtbsrHashAlgorithmOID: String? = null,
+) {
+    /**
+     * If in each of the following bullet points one of the mentioned parameters is
+     * present, the other must be present:
+     *  “hash” and “hashAlgorithmOID”
+     *  “documentLocation_uri” and “documentLocation_method”
+     *  “dtbsr” and “dtbsrHashAlgorithmOID”
+     * In each of the following bullet points at least one of the mentioned
+     * parameters must be present:
+     *  “hash” or “dtbsr”
+     */
+    init {
+        require(hash != null || dtbsr != null)
+        require(hashAlgorithmOID iff hash)
+        require(dtbsrHashAlgorithmOID iff dtbsr)
+        require(documentLocationUri?.toString() iff hash)
+        require(documentLocationMethod?.toString() iff documentLocationUri?.toString())
+    }
+
+    @Serializable
+    @SerialName("documentLocation_method")
+    data class DocumentLocationMethod private constructor(
+        val method: Method,
+        val oneTimePassword: String? = null,
+    ) {
+        /**
+         * If “document_access_mode” is “OTP”, “oneTimePassword” must be
+         * present.
+         */
+        init {
+            require(
+                (oneTimePassword == null && method != Method.OTP)
+                        || (oneTimePassword != null && method == Method.OTP)
+            )
+        }
+
+        /**
+         * this is potentially a mistake in the draft spec vs test vector,
+         * currently we need it to be a sealed class with polymorphic serialization to get the structure
+         * method: {type: NAME}
```

Comment:
Please use three backticks to mark a source block

Reply:
Using Preview there is no difference between one and three backticks. Is it okay now?